cucumber_cpp/library/cucumber_expression/Ast.hpp

```diff
@@ -0,0 +1,89 @@
+#ifndef CUCUMBER_EXPRESSION_AST_HPP
+#define CUCUMBER_EXPRESSION_AST_HPP
+
+#include <cstddef>
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace cucumber_cpp::library::cucumber_expression
+{
+    enum class NodeType
+    {
+        text,
+        optional,
+        alternation,
+        alternative,
+        parameter,
+        expression,
+    };
+
+    enum class TokenType
+    {
+        startOfLine,
+        endOfLine,
+        whiteSpace,
+        beginOptional,
+        endOptional,
+        beginParameter,
+        endParameter,
+        alternation,
+        text,
+        invalid,
+    };
+
+    struct Node
+    {
+        Node(NodeType type,
+            std::size_t start,
+            std::size_t end,
+            std::variant<std::string, std::vector<Node>> children);
+
+        NodeType Type() const;
+        std::size_t Start() const;
+        std::size_t End() const;
+
+        std::string Text();
+        std::string Text() const;
+
+        std::vector<Node>& Children();
+        const std::vector<Node>& Children() const;
+
+        const std::variant<std::string, std::vector<Node>>& GetLeafNodes() const;
+
+        bool operator==(const Node& other) const = default;
```

Comment:
If you provide operator==, then please also provide operator!= for symmetry

Reply:
this is a c++20 thing and completely valid and recommended